src/test/csfle.rs

```diff
@@ -2873,6 +2881,376 @@ async fn bypass_mongocryptd_client() -> Result<()> {
+        .await?
+        .try_collect()
+        .await?;
+    assert_success(docs, &[200]);
+
+    // Case 5: Run an aggregation expression inside $expr
+    let query = rawdoc! { "$and": [ { "$lt": [ format!("${key}"), get_raw_bson_from_num(bson_type, 30) ] } ] };
+    let find_payload = client_encryption
+        .encrypt_expression(query, key1_id.clone())
+        .contention_factor(0)
+        .range_options(range_options.clone())
+        .run()
+        .await?;
+
+    let docs: Vec<RawDocumentBuf> = encrypted_client
+        .database("db")
+        .collection("explicit_encryption")
+        .find(doc! { "$expr": find_payload }, find_options.clone())
+        .await?
+        .try_collect()
+        .await?;
+    assert_success(docs, &[0, 6]);
+
+    // Case 6: Encrypting a document greater than the maximum errors
+    if bson_type != "DoubleNoPrecision" && bson_type != "DecimalNoPrecision" {
+        let num = get_raw_bson_from_num(bson_type, 201);
+        let error = client_encryption
+            .encrypt(num, key1_id.clone(), Algorithm::RangePreview)
+            .contention_factor(0)
+            .range_options(range_options.clone())
+            .run()
+            .await
+            .unwrap_err();
+        assert!(matches!(*error.kind, ErrorKind::Encryption(_)));
+    }
+
+    // Case 7: Encrypting a document of a different type errors
+    if bson_type != "DoubleNoPrecision" && bson_type != "DecimalNoPrecision" {
+        let value = if bson_type == "Int" {
+            rawdoc! { &key: { "$numberDouble": "6" } }
+        } else {
+            rawdoc! { &key: { "$numberInt": "6" } }
+        };
+        let error = client_encryption
+            .encrypt(value, key1_id.clone(), Algorithm::RangePreview)
+            .contention_factor(0)
+            .range_options(range_options.clone())
+            .run()
+            .await
+            .unwrap_err();
+        assert!(matches!(*error.kind, ErrorKind::Encryption(_)));
+    }
+
+    // Case 8: Setting precision errors if the type is not a double
+    if !bson_type.contains("Double") && !bson_type.contains("Decimal") {
+        let range_options = RangeOptions::builder()
+            .sparsity(1)
+            .min(get_bson_from_num(bson_type, 0))
+            .max(get_bson_from_num(bson_type, 200))
+            .precision(2)
+            .build();
+        let error = client_encryption
+            .encrypt(
+                bson_numbers[&6].clone(),
+                key1_id.clone(),
+                Algorithm::RangePreview,
+            )
+            .contention_factor(0)
+            .range_options(range_options)
+            .run()
+            .await
+            .unwrap_err();
+        assert!(matches!(*error.kind, ErrorKind::Encryption(_)));
+    }
+
+    Ok(())
+}
+
+fn get_bson_from_num(bson_type: &str, num: i32) -> Bson {
+    match bson_type {
+        "DecimalNoPrecision" | "DecimalPrecision" => {
+            Bson::Decimal128(num.to_string().parse().unwrap())
```

Comment:
I think this is the best way to get from a number to a Decimal128 right now? It might be worth adding some other utility functions to avoid intermediary strings here.

Reply:
Makes sense, filed RUST-1619 to track that.